File: IFATCCompanion/Weather/SIGMETParser.swift
import Foundation
import CoreLocation

/// Parses AIRMET/SIGMET/G-AIRMET hazard products from the Aviation Weather Center
/// JSON API (best-effort; coverage and schema vary by product).
enum SIGMETParser {

    static func parseJSON(_ data: Data) -> [SIGMET] {
        JSONLenient.array(data).compactMap { obj in
            var s = SIGMET(raw: JSONLenient.string(obj["rawSigmet"]) ?? JSONLenient.string(obj["rawAirSigmet"]) ?? JSONLenient.string(obj["raw_text"]) ?? "")
            s.hazard = JSONLenient.string(obj["hazard"]) ?? JSONLenient.string(obj["hazardType"])
            s.severity = JSONLenient.string(obj["severity"])
            if let coords = obj["coords"] as? [[String: Any]] {
                s.area = coords.compactMap { c in
                    guard let lat = JSONLenient.double(c["lat"]), let lon = JSONLenient.double(c["lon"]) else { return nil }
                    return CLLocationCoordinate2D(latitude: lat, longitude: lon)
                }
            }
            if s.raw.isEmpty && s.hazard == nil { return nil }
            return s
        }
    }
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/weather/SigmetParser.kt
package com.h3consultingpartners.ifatccompanion.core.weather

import com.h3consultingpartners.ifatccompanion.core.geo.Coordinate

/**
 * Parses AIRMET/SIGMET/G-AIRMET hazard products from the Aviation Weather Center
 * JSON API (best-effort; coverage and schema vary by product).
 *
 * Ported from `IFATCCompanion/Weather/SIGMETParser.swift`.
 */
object SigmetParser {

    fun parseJson(data: ByteArray): List<SIGMET> = JsonLenient.array(data).mapNotNull { obj ->
        val s = SIGMET(
            raw = JsonLenient.string(obj["rawSigmet"])
                ?: JsonLenient.string(obj["rawAirSigmet"])
                ?: JsonLenient.string(obj["raw_text"])
                ?: "",
        )
        s.hazard = JsonLenient.string(obj["hazard"]) ?: JsonLenient.string(obj["hazardType"])
        s.severity = JsonLenient.string(obj["severity"])
        JsonLenient.objectArray(obj["coords"])?.let { coords ->
            s.area = coords.mapNotNull { c ->
                val lat = JsonLenient.double(c["lat"]) ?: return@mapNotNull null
                val lon = JsonLenient.double(c["lon"]) ?: return@mapNotNull null
                Coordinate(latitude = lat, longitude = lon)
            }
        }
        if (s.raw.isEmpty() && s.hazard == null) return@mapNotNull null
        s
    }
}
